ENG-622 - 1.60.0 - OIDC prompt parameter (#3891)
FusionAuth/fusionauth-site: astro/src/content/docs/lifecycle/authenticate-users/oauth/prompt.mdx

##### Third party identity providers

The `prompt` parameter will not be forwarded to third party identity providers when clicking a "Login with" buton on the login page, or when using the `idp_hint` parameter, or a `login_hint` parameter with an IdP configured to manage the corresponding email domain that would result in FusionAuth bypassing the login page and redirecting to the third party identity provider.

The reason for this decision is because this is an optional parameter in the OpenID specification, it is not known if every OpenID Connect identity provider will honor this parameter, and it is unlikely that other supported identity providers that are not OpenID Connect compliant will honor this parameter. Additionally and most important, it may cause integration failures due to unexpected parameters being sent to third party identity providers.

If you require the ability to optionally forward the `prompt` parameter to a third party identity provider, please open a GitHub issue, or contact your account manager. We would love to receive your feedback.

#### Edge cases

> One edge case to note is that if begin a request with `idp_hint` and `prompt=login` FusionAuth will still redirect to the identity provider request by the `idp_hint`. The reason for this is generally speaking when `idp_hint` is provided, you are telling FusionAuth this user must be authenticated by a third party, do not show the login page. In this case it would not be useful to then show the user the FusionAuth login page knowing the cannot authenticate. If you encounter this edge case, you should likely just remove the `prompt=login` from the request. The same behavior will occur when using `login_hint` with managed domains. This may be improved in the future by allowing for configuration to indicate if it is safe to forward the `prompt` parameter to the third party.

Reviewer: ```suggestion
One edge case to note is that if a request begins with `idp_hint` and `prompt=login`, FusionAuth will still redirect to the identity provider request by the `idp_hint`. The reason for this is generally speaking when `idp_hint` is provided, you are telling FusionAuth this user must be authenticated by a third party, do not show the login page. In this case it would not be useful to then show the user the FusionAuth login page knowing the cannot authenticate. If you encounter this edge case, you should likely just remove the `prompt=login` from the request. The same behavior will occur when using `login_hint` with managed domains. This may be improved in the future by allowing for configuration to indicate if it is safe to forward the `prompt` parameter to the third party.
```

Author: Thanks, I'll rework this, but I don't want to say begins with, that isn't exactly what I'm trying to get across either.

#### Consent
The `consent` prompt only applies when FusionAuth has otherwise been configured to display consent prompts. For example, if the application is not configured as a third party application, or you have configured the consent mode to always prompt or never prompt, adding `consent` to the prompt parameter will have no effect.`

> When the consent mode is remember decision, then adding `consent` to the prompt parameter will cause the user to be prompted again regardless of their pior decision.

Reviewer: ```suggestion
When the consent mode is remember decision, then adding `consent` to the prompt parameter will cause the user to be prompted again regardless of their prior decision.
```

Author: Thanks! I'll see if I made this typo anywhere else as well.

- `error=consent_required`
- `error_reason=consent_required`
- `error_descriotion=The user is required to consent to one or more scopes in order to complete authentication.`

See [OAuth2 Error Redirect parameters](/docs/lifecycle/authenticate-users/oauth/endpoints#redirect-parameters-2) for additional information.

### Security considerations
Note that you should use `prompt=login` or `max_age` with the correct expectations. Request parameters can be easily removed from the request, so you should not rely upon these parameters for any level of security or guarantee that the user has been reauthenticated.

If a user were to remove the `prompt=login` parameter and the user had an existing SSO session, then existing session would be used, an authorization code would be generated and redirected back to the authorized redirect.

The prompt parameter is intended to be a suggestion to the identity provider but by itself provides no guarantee to the relying party (RP) that the request was honored, and the user has been reauthenticated.

> If you are using `prompt=login` for security purposes in order to ensure the user has recently authentication, you must inspect and validate the `auth_time` claim in the token you are using whether it is the `id_token` or `access_token`.

Reviewer 2: ```suggestion
To use `prompt=login` to ensure the user authenticated recently, you must also inspect and validate the `auth_time` claim in the appropriate token, either the `id_token` or `access_token`.
```

The `auth_time` claim represents the time of the actual user authentication and is expressed as [unix time](/docs/reference/data-types#unix-time).

Using this value, you can assert when the last user authenticate event occurred and as a result know that `prompt=login` was honored by the identity provider and was not removed by a malicious user.

> For example, when you utilize `login=prompt` or `max_age`, when you complete the auth code exchange when using the authorization code grant you will inspect the `auth_time` and compare it to the current token you have to assert the new token has a more recent `auth_time` claim, and that the value is within your expected time frame.

Reviewer 2: ```suggestion
For example, when using `login=prompt` or `max_age`, after completing the authorization code grant auth code exchange compare the `auth_time` to the current token. The new token must have a more recent `auth_time` claim and the value must be within your expected time frame.
```

> If the `auth_time` has not changed from your current access token, or for whatever reason is not recent enough - you may reject the token and end the users session, or fail to allow them to perform the requested task that initiated this login request.

Reviewer 2: ```suggestion
If `auth_time` is the same as your current access token, or isn't recent enough, reject the token and end the users session. Alternatively, don't allow the user to perform the requested task that initiated the login request.
```

> When using `prompt=login` or `max_age=` parameters, you must also be inspecting the `auth_time` claim to prove the user has authenticated with a time frame that meets your security profile.

Reviewer 2: As far as I can tell this line is a repeat of what's already been said in at least three ways... Remove it.

### Limitations and constraints

> The following are some known limitations and constraints that may be useful to understand as you begin using this new parameter.

Reviewer 2: ```suggestion
Before using this parameter make sure you understand the following are some known limitations and constraints.
```

#### Unsupported values

> The OpenID connect specification also defines `select_account`. This parameter is not yet supported. Providing this or any additional values in the `prompt` parameter will be ignored.

Reviewer 2: ```suggestion
The Open ID `select_account` parameter is currently unsupported. Using this value in `prompt` is ignored.
```

##### Third party identity providers

> The `prompt` parameter will not be forwarded to third party identity providers when clicking a "Login with" buton on the login page, or when using the `idp_hint` parameter, or a `login_hint` parameter with an IdP configured to manage the corresponding email domain that would result in FusionAuth bypassing the login page and redirecting to the third party identity provider.

Reviewer 2: ```suggestion
The `prompt` parameter isn't forwarded to third party identity providers in the following cases:
- Clicking a "Login with" button on the login page.
- Using the `idp_hint` parameter, or a `login_hint` parameter with an IdP that is configured to manage the corresponding email domain that would result in FusionAuth bypassing the login page and redirecting to the third party identity provider.
```

> The reason for this decision is because `prompt` is an optional parameter in the OpenID specification and we cannot know if every OpenID Connect identity provider will honor this parameter. Additionally FusionAuth supports many identity providers, many of which are actually OAuth2, or partial implementations of OAuth2 and OpenID Connect. If we consider all supported identity providers, it is unlikely that all will honor this parameter, and more importantly sending unknown or unexpected parameters to some identity providers will cause a failure.

Reviewer 2: ```suggestion
This is because `prompt` is an optional parameter in the OpenID specification so may not be supported by every identity provider. FusionAuth supports many identity providers, some of which are OAuth2, or partial implementations of OAuth2 and OpenID Connect. Sending unknown or unexpected parameters to some of those identity providers causes a failure.
```

> For these reasons, we decided to not forward the `prompt` parameter for now. This may be improved in the future by allowing for configuration to indicate if the third party identity provider supports the `prompt` parameter, and which prompt values you would like to forward. It may even be worth while to allow for this value to be translated if the third party supports similar behavior but with a differently named parameter. For example, SAML v2 has a parameter named `ForceAuthN` which provides a similar function to `prompt=login`.

Reviewer 2: This is a classic "forward looking statement" that can easily be overridden by higher priorities but frustrate the heck out of customers... delete this paragraph as there's nothing a developer can do with the information.

If you require the ability to optionally forward the `prompt` parameter to a third party identity provider, please open a GitHub issue, or contact your account manager. We would love to receive your feedback.

#### Edge cases

> If you are using the `idp_hint` request parameter and also include `prompt=login` FusionAuth will still redirect to the identity provider resolved by request parameter.

Reviewer 2: ```suggestion
When using the `idp_hint` request parameter and including `prompt=login`, FusionAuth redirects to the identity provider resolved by request parameter.
```

> The reason for this decision is that when the `idp_hint` is provided, you have provided explicit instruction to FusionAuth that this user must be authenticated by a third party and often the caller does not want the user to see the FusionAuth login page. In this context, it would not be useful to then show the user the FusionAuth login page knowing FusionAuth will not be able to authenticate the user. It is possible this is not the desired behavior for all customers.

Reviewer 2: ```suggestion
The `idp_hint` indicates that this user must be authenticated by a third party and often the caller does not want the user to see the FusionAuth login page. In this context, it's not useful to show the user the FusionAuth login page as FusionAuth is unable to authenticate the user. It is possible this is not the desired behavior for all customers.
```

> If you encounter this edge case, you should remove the `prompt=login` from the request. Using a `login_hint` request parameter with managed domains will result in the same behavior.  This may be improved in the future by allowing for additional identity provider configuration for prompt handling, see the above section on third party identity providers for additional details.

Reviewer 2: ```suggestion
If this behavior happens, remove the `prompt=login` from the request. Using a `login_hint` request parameter with managed domains results in the same behavior.  This may be improved in the future by allowing for additional identity provider configuration for prompt handling, see the above section on third party identity providers for additional details.
```
